pkg/aws/metadata/metadata.go

```diff
@@ -11,53 +11,78 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
 package metadata
 
 import (
+	"context"
 	"fmt"
 	"io"
-	"net/http"
+	"log"
+
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/feature/ec2/imds"
 )
 
-func getMetadata(name string) (string, error) {
-	resp, err := http.Get("http://169.254.169.254/latest/meta-data/" + name)
+func newClient() (*imds.Client, error) {
+	cfg, err := config.LoadDefaultConfig(context.TODO())
+	if err != nil {
+		log.Printf("error: %v", err)
+		return nil, err
+	}
+
+	client := imds.NewFromConfig(cfg)
+
+	return client, err
+}
+
+func getMetadata(client *imds.Client, path string) (string, error) {
+	res, err := client.GetMetadata(context.TODO(), &imds.GetMetadataInput{
+		Path: path,
+	})
 	if err != nil {
-		return "", fmt.Errorf("unable to retrieve instance-id from metadata server: %s", err)
+		return "", err
 	}
 
-	defer resp.Body.Close()
-	instanceID, err := io.ReadAll(resp.Body)
+	defer res.Content.Close()
+	value, err := io.ReadAll(res.Content)
 	if err != nil {
-		return "", fmt.Errorf("unable to read response body: %s", err)
+		return "", err
 	}
 
-	return string(instanceID), nil
+	return string(value), err
 }
 
-// GetInstanceMetadata returns the instance ID and type
+// GetInstanceMetadata returns required AWS metadatas
 func GetInstanceMetadata() (instanceID, instanceType, availabilityZone, vpcID string, err error) {
-	instanceID, err = getMetadata("instance-id")
+	client, err := newClient()
```

Comment:
Hmm, should we be creating a new client every time there's a call to `GetInstanceMetadata()`? Does it not matter?

Reply:
Thanks @christarazi for the review. regarding the client creation I don't see where else should I put it.
I checked the [nodediscovery.go](https://github.com/cilium/cilium/blob/master/pkg/nodediscovery/nodediscovery.go) which make use of this function and it seems to be called once. Please let me know if there's a better place

Reply:
Fair enough, I think if it's called once, then the code is fine as-is